File: src/src_core/fmg/core/img/BurgerMenuImg.java
package fmg.core.img;

import java.util.stream.Stream;

import fmg.common.Color;
import fmg.common.HSV;
import fmg.common.geom.PointDouble;
import fmg.common.geom.RectDouble;
import fmg.common.geom.util.FigureHelper;

/**
 * Abstract representable menu as horizontal or vertical lines
 * @param <TImage> plaform specific image
 */
public abstract class BurgerMenuImg<TImage> extends PolarLightsImg<TImage> {

   protected BurgerMenuImg() {
      _showBurgerMenu = true;
      _layersInBurgerMenu = 3;
      _horizontalBurgerMenu = true;
      _rotateBurgerMenu = true;
   }

   public static final String PROPERTY_SHOW_BURGER_MENU = "ShowBurgerMenu";
   public static final String PROPERTY_HORIZONTAL_BURGER_MENU = "HorizontalBurgerMenu";
   public static final String PROPERTY_LAYERS_IN_BURGER_MENU = "LayersInBurgerMenu";
   public static final String PROPERTY_ROTATE_BURGER_MENU = "RotateBurgerMenu";

   private boolean _showBurgerMenu;
   public boolean isShowBurgerMenu() { return _showBurgerMenu; }
   public void   setShowBurgerMenu(boolean value) { setProperty(_showBurgerMenu, value, PROPERTY_SHOW_BURGER_MENU); }

   private boolean _horizontalBurgerMenu;
   public boolean isHorizontalBurgerMenu() { return _horizontalBurgerMenu; }
   public void   setHorizontalBurgerMenu(boolean value) { setProperty(_horizontalBurgerMenu, value, PROPERTY_HORIZONTAL_BURGER_MENU); }

   private int   _layersInBurgerMenu;
   public int  getLayersInBurgerMenu() { return _layersInBurgerMenu; }
   public void setLayersInBurgerMenu(int value) { setProperty(_layersInBurgerMenu, value, PROPERTY_LAYERS_IN_BURGER_MENU); }

   private boolean _rotateBurgerMenu;
   public boolean isRotateBurgerMenu() { return _rotateBurgerMenu; }
   public void   setRotateBurgerMenu(boolean value) { setProperty(_rotateBurgerMenu, value, PROPERTY_ROTATE_BURGER_MENU); }

   protected RectDouble getBurgerMenuRegion() {
      return new RectDouble(getWidth() /2.0,
                            getHeight()/2.0,
                            getWidth() /2.0 - getPadding().right,
                            getHeight()/2.0 - getPadding().bottom);
   }

   protected static class LineInfo {
      public Color clr;
      public double penWidht;
      public PointDouble from; // start coord
      public PointDouble to;   // end   coord
   }

   protected Stream<LineInfo> getCoordsBurgerMenu() {
      if (!isShowBurgerMenu())
         return Stream.empty();

      boolean horizontalMenu = isHorizontalBurgerMenu();
      int layers = getLayersInBurgerMenu();
      RectDouble rcMenu = getBurgerMenuRegion();
      double penWidth = Math.max(1, (horizontalMenu ? rcMenu.height : rcMenu.width) / (2 * layers));
      double angle = isRotateBurgerMenu() ? getRotateAngle() : 0;
      double stepAngle = 360.0 / layers;
      return FigureHelper.getBurgerMenu(rcMenu, layers, angle, horizontalMenu)
         .map(t -> {
            int layerNum = t.first;
            HSV hsv = new HSV(Color.Gray);
            hsv.v *= Math.sin(layerNum*stepAngle / layers);

            LineInfo li = new LineInfo();
            li.clr = hsv.toColor();
            li.penWidht = penWidth;
            li.from = t.second.first;
            li.to = t.second.second;
            return li;
         });
   }

}

File: src/src_core/fmg/core/img/BurgerMenuImg.cs
using System;
using System.Linq;
using System.Collections.Generic;
using fmg.common;
using fmg.common.geom;

namespace fmg.core.img {

   /// <summary>
   /// Abstract representable menu as horizontal or vertical lines
   /// </summary>
   /// <typeparam name="TImage">plaform specific image</typeparam>
   public abstract class BurgerMenuImg<TImage> : PolarLightsImg<TImage>
      where TImage : class
   {
      protected BurgerMenuImg() {
         _showBurgerMenu = true;
         _layersInBurgerMenu = 3;
         _horizontalBurgerMenu = true;
         _rotateBurgerMenu = true;
      }

      private bool _showBurgerMenu;
      public bool ShowBurgerMenu {
         get { return _showBurgerMenu; }
         set {
            if (SetProperty(ref _showBurgerMenu, value)) {
               Invalidate();
            }
         }
      }

      private bool _horizontalBurgerMenu;
      public bool HorizontalBurgerMenu {
         get { return _horizontalBurgerMenu; }
         set {
            if (SetProperty(ref _horizontalBurgerMenu, value)) {
               Invalidate();
            }
         }
      }

      private int _layersInBurgerMenu;
      public int LayersInBurgerMenu {
         get { return _layersInBurgerMenu; }
         set {
            if (SetProperty(ref _layersInBurgerMenu, value)) {
               Invalidate();
            }
         }
      }

      private bool _rotateBurgerMenu;
      public bool RotateBurgerMenu {
         get { return _rotateBurgerMenu; }
         set {
            if (SetProperty(ref _rotateBurgerMenu, value)) {
               Invalidate();
            }
         }
      }

      private Bound? _paddingBurgerMenu;
      public Bound PaddingBurgerMenu {
         get {
            if (_paddingBurgerMenu == null)
               // call this setter
               PaddingBurgerMenu = new Bound(Size.Width / 2,
                                             Size.Height / 2,
                                             Padding.Right,
                                             Padding.Bottom);
            return _paddingBurgerMenu.Value;
         }
         set {
            if (value.LeftAndRight >= Size.Width)
               throw new ArgumentException("Padding size is very large. Should be less than Width.");
            if (value.TopAndBottom >= Size.Height)
               throw new ArgumentException("Padding size is very large. Should be less than Height.");
            if (SetProperty(ref _paddingBurgerMenu, value)) {
               Invalidate();
            }
         }
      }
      public void ResetPaddingBurgerMenu() {
         if (_paddingBurgerMenu == null)
            return;
         _paddingBurgerMenu = null;
         OnPropertyChanged(nameof(this.PaddingBurgerMenu));
         Invalidate();
      }

      protected struct LineInfo {
         public Color clr;
         public double penWidht;
         public PointDouble from; // start coord
         public PointDouble to;   // end   coord
      }

      protected IEnumerable<LineInfo> GetCoordsBurgerMenu() {
         if (!ShowBurgerMenu)
            return Enumerable.Empty<LineInfo>();

         bool horizontal = HorizontalBurgerMenu;
         int layers = LayersInBurgerMenu;
         var pad = PaddingBurgerMenu;
         Rect rc = new Rect(pad.Left,
                            pad.Top,
                            Size.Width  - pad.LeftAndRight,
                            Size.Height - pad.TopAndBottom);;
         double penWidth = Math.Max(1, (horizontal ? rc.Height : rc.Width) / (2.0 * layers));
         double rotateAngle = RotateBurgerMenu ? RotateAngle : 0;
         double stepAngle = 360.0 / layers;

         return Enumerable.Range(0, layers)
            .Select(layerNum => {
               double layerAlignmentAngle = FixAngle(layerNum * stepAngle + rotateAngle);
               double offsetTop = !horizontal ? 0 : layerAlignmentAngle * rc.Height / 360;
               double offsetLeft = horizontal ? 0 : layerAlignmentAngle * rc.Width / 360;
               PointDouble start = new PointDouble(rc.Left() + offsetLeft,
                                                   rc.Top() + offsetTop);
               PointDouble end = new PointDouble((horizontal ? rc.Right() : rc.Left()  ) + offsetLeft,
                                                 (horizontal ? rc.Top()   : rc.Bottom()) + offsetTop);

               HSV hsv = new HSV(Color.Gray);
               hsv.v *= Math.Sin(layerNum * stepAngle / layers);

               LineInfo li = new LineInfo();
               li.clr = hsv.ToColor();
               li.penWidht = penWidth;
               li.from = start;
               li.to = end;
               return li;
            });
      }

   }
}
